internal/artifacts/module_arch.go

```diff
@@ -93,18 +96,57 @@ func ExecuteSoloBuild(source, target string, extensions []string, modulesNames [
 		logs.Logger.Infof(multiBuildMsg, `"`+strings.Join(sortedModules, `", "`)+`"`)
 	}
 
-	err = buildModules(sourceDir, target, extensions, sortedModules, selectedModulesMap, wdGetter)
+	packedModules, err := buildModules(sourceDir, target, extensions, sortedModules, selectedModulesMap, wdGetter)
 	if err != nil {
 		return wrapBuildError(err, modulesNames)
 	}
 
+	if generateMtad {
+		// validate platform
+		platform, err := validatePlatform(platform)
+		if err != nil {
+			return wrapBuildError(err, modulesNames)
+		}
+
+		err = removeBuildParamsFromMta(loc, mtaObj, false)
```

Comment:
Please extract a function that converts the mta to mtad.yaml and generates the mtad instead of copying its logic.
This will probably also let you remove the `marshal` function parameter (you can send the correct `Marshal` when calling this function, and the error handling code will be covered by other errors like the call to `wdGetter`).